File: frontend/lib/utils/mrz_utils.dart
class MRZUtils {
  /// Calculates the check digit for a string according to ICAO 9303 standards.
  /// It uses the 7-3-1 weighting system.
  static int calculateChecksum(String data) {
    const weights = [7, 3, 1];
    int sum = 0;

    for (int i = 0; i < data.length; i++) {
      final codeUnit = data.codeUnitAt(i);
      int val = 0;

      if (codeUnit >= 48 && codeUnit <= 57) {
        // '0'-'9'
        val = codeUnit - 48;
      } else if (codeUnit >= 65 && codeUnit <= 90) {
        // 'A'-'Z'
        val = codeUnit - 65 + 10;
      } else {
        // '<' and other fillers are 0
        val = 0;
      }

      sum += val * weights[i % 3];
    }

    return sum % 10;
  }
}

File: back-end/pkg/nfc/mrz_utils.go
package nfc

// CalculateMRZChecksum calculates the check digit for a string according to ICAO 9303 standards.
// It uses the 7-3-1 weighting system.
func CalculateMRZChecksum(data string) int {
	weights := [3]int{7, 3, 1}
	sum := 0

	// Optimize: iterate byte-by-byte over ASCII string instead of UTF-8 rune decoding
	for i := 0; i < len(data); i++ {
		char := data[i]
		var val int
		switch {
		case char >= '0' && char <= '9':
			val = int(char - '0')
		case char >= 'A' && char <= 'Z':
			val = int(char - 'A' + 10)
		default:
			// '<' and other fillers are 0
			val = 0
		}

		sum += val * weights[i%3]
	}

	return sum % 10
}
